### PR TITLE
BUG: Fix min_count issue for groupby.sum

### LINKED ISSUE #32861: Calling `sum` with `min_count` on `SeriesGroupBy` with dtype `Int64` gives large negative value rather than pd.NA
```python

test_df = pd.DataFrame({'foo' : ['a'], 'bar': [1]})
test_df['bar'] = test_df['bar'].astype('Int64')
test_df.groupby('foo')['bar'].sum(min_count=2)

# output:
foo
a    -9223372036854775808
Name: bar, dtype: Int64

```
#### Problem description

Per the documentation, `sum` should return `NA` if there are fewer than `min_count` values. This works fine on the dataframe itself:

```python
test_df['bar'].sum(min_count=2)

# output
nan
```

but gives what looks like an overflow error when called after `groupby`.

I ran into this with real data when calling with `min_count=1` on a dataframe where some of the values were missing, but I thought the minimal example above was clearer.

#### Expected Output

foo
a    NaN
Name: bar, dtype: Int64

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-74-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2019.1
dateutil         : 2.8.0
pip              : 19.1.1
setuptools       : 41.0.1
Cython           : None
pytest           : 5.3.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.3.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.5.0
pandas_datareader: None
bs4              : 4.8.1
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.3.3
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.3.4
pyxlsb           : None
s3fs             : None
scipy            : 1.3.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #32861
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_function.py [hunk 1]
```diff
@@ -1636,3 +1636,20 @@ def test_apply_to_nullable_integer_returns_float(values, function):
...
+    grouped = df.groupby("a")
+    idx = pd.Index([0, 1, 2], dtype=object, name="a")
+
+    result = grouped["b"].sum(min_count=2)
+    expected = pd.Series([np.nan] * 3, index=idx, name="b")
+    tm.assert_series_equal(result, expected)
+
+    result = grouped.sum(min_count=2)
+    expected = pd.DataFrame(
+        {"b": [pd.NA] * 3, "c": [pd.NA] * 3}, dtype="Int64", index=idx
```
Comment (dsaxton) [inline pandas/tests/groupby/test_function.py]: Not ideal but it seems we get `NA` in the `DataFrame` case but `NaN` for `Series`. Should I add a FIXME comment with a follow-up issue?

File: pandas/core/groupby/ops.py [hunk 2]
```diff
@@ -553,7 +553,8 @@ def _cython_operation(
             # Two options for avoiding this special case
             # 1. mask-aware ops and avoid casting to float with NaN above
             # 2. specify the result dtype when calling this method
-            result = result.astype("int64")
+            if not isna(result).any():
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: move this condition to the above elif & add an appropriate comment as 3.

File: pandas/tests/groupby/test_function.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/groupby/test_function.py]: huh? that seems odd, can you track this down

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py [hunk 3]
```diff
@@ -577,6 +583,14 @@ def _cython_operation(
         elif is_datetimelike and kind == "aggregate":
             result = result.astype(orig_values.dtype)
 
+        if (
```
Comment (dsaxton) [inline pandas/core/groupby/ops.py]: This isn't pretty but don't really know of a better way to make SeriesGroupBy return NA instead of NaN

Comment (jreback) [reply pandas/core/groupby/ops.py]: see #32984;
after we merge that you can just make a small mod to use the same

Comment (jreback) [reply pandas/core/groupby/ops.py]: merged #32894 if you want to refactor

Comment (jreback) [reply pandas/core/groupby/ops.py]: you need to use maybe_cast_result_dtype here; I don't actually think we need the else any longer

File: doc/source/whatsnew/v1.1.0.rst [hunk 4]
```diff
@@ -400,8 +400,6 @@ Groupby/resample/rolling
 
 - Bug in :meth:`GroupBy.apply` raises ``ValueError`` when the ``by`` axis is not sorted and has duplicates and the applied ``func`` does not mutate passed in objects (:issue:`30667`)
 - Bug in :meth:`DataFrameGroupby.transform` produces incorrect result with transformation functions (:issue:`30918`)
-- Bug in :meth:`DataFrame.groupby` and :meth:`Series.groupby` produces inconsistent type when aggregating Boolean series (:issue:`32894`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: you removed some things

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 3 above)
Comment (dsaxton) [reply pandas/core/groupby/ops.py]: A little confused, do you mean `maybe_cast_result` (`maybe_cast_result_dtype` returns a numpy int dtype itself)? When I try using `maybe_cast_result` though I just get a numpy array back

Comment (jreback) [reply pandas/core/groupby/ops.py]: look at the function which takes the name of the operation 'add' and the dtype and gives you back the casting dtype. which is eactly what we need to do here; you just need to modfy maybe_cast_result.

rather than have an else at all.

Comment (dsaxton) [reply pandas/core/groupby/ops.py]: This seems to be causing lots of test failures (e.g., also sending numpy arrays down this path raises when maybe_cast_result tries to access a _values attribute). Also seems we'd need to remove the check for result[0] being an instance of the original dtype because then maybe_cast_result doesn't try to convert NaN to a nullable integer, but then categorical tests end up failing. Seems worth refactoring but I'm honestly not sure how.

Comment (jreback) [reply pandas/core/groupby/ops.py]: i would like to fine a way of using maybe_cast_result, i don't mind have a very simpl elif here. but this is not good as written.

Comment (dsaxton) [reply pandas/core/groupby/ops.py]: Made a couple changes (some from another PR) that simplifies it a little

File: pandas/core/groupby/ops.py [hunk 5]
```diff
@@ -575,6 +565,11 @@ def _cython_operation(
         elif is_datetimelike and kind == "aggregate":
             result = result.astype(orig_values.dtype)
 
+        if is_integer_dtype(orig_values.dtype) and is_extension_array_dtype(
```
Comment (jreback) [inline pandas/core/groupby/ops.py:568]: the whole entire point of `maybe_cast_result` is that you don't need the if here

to be honest we should remove lines 558-566, but can do that in another PR

File: pandas/core/dtypes/cast.py [hunk 6]
```diff
@@ -275,12 +276,13 @@ def maybe_cast_result(
...
-            # type if it's compatible.
-            if len(result) and isinstance(result[0], dtype.type):
-                cls = dtype.construct_array_type()
-                result = maybe_cast_to_extension_array(cls, result, dtype=dtype)
+        if (
+            is_extension_array_dtype(dtype)
+            and dtype.kind != "M"
+            and not is_categorical_dtype(dtype)
+        ):
+            cls = dtype.construct_array_type()
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:284]: IF you need extra logic, then you can add it here.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 5 above)
Comment (dsaxton) [reply pandas/core/groupby/ops.py]: Without the if about 550 groupby tests fail, though I think it works when only checking for extension arrays

File: pandas/core/groupby/ops.py [hunk 7]
```diff
@@ -575,6 +565,9 @@ def _cython_operation(
         elif is_datetimelike and kind == "aggregate":
             result = result.astype(orig_values.dtype)
 
+        if is_extension_array_dtype(orig_values.dtype):
```
Comment (jreback) [inline pandas/core/groupby/ops.py:568]: if you move line 568 (the is_extension_array_dtype) before line 558, can we then completely remove lines 558-556? (these are *already* extension dtypes).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (dsaxton) [reply pandas/core/groupby/ops.py:568]: No, it looks like we still get a lot of failures

**Comment by simonjayhawkins** [comment]
@dsaxton what's the status here?

**Comment by dsaxton** [comment]
> @dsaxton what's the status here?

It's working / green, but I think @jreback may want it refactored (I haven't had much success in deleting what looks like redundant code without causing tests to fail)

**Comment by simonjayhawkins** [comment]
The 'regression' for Int64 could be considered a bug since Int64 is considered experimental. 

I've not looked in detail yet, but the 'regression' was caused by #31359 see https://github.com/pandas-dev/pandas/issues/32861#issuecomment-605453895 so would to check that the other EAs which aren't experimental aren't affected by this issue.


> but I think @jreback may want it refactored

if this issue **does** affect other EAs, then this PR is a candidate for backport, so would rather keep the changes here minimal.

**Comment by dsaxton** [comment]
> **if** this issue **does** affect other EAs, then this PR is a candidate for backport, so would rather keep the changes here minimal.

It doesn't affect others that I'm aware of

**Comment by dsaxton** [comment]
@jreback @simonjayhawkins Is is possible we can merge this as is? I have a fix for a separate bug that depends on this.

**Comment by jreback** [comment]
> @jreback @simonjayhawkins Is is possible we can merge this as is? I have a fix for a separate bug that depends on this.

sure is this rebase on master?

File: pandas/tests/groupby/test_function.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/groupby/test_function.py:1656]: can you create an issue about this. I am not sure this is correct.

**Comment by dsaxton** [comment]
> sure is this rebase on master?

I think Simon merged this morning

File: pandas/tests/groupby/test_function.py (hunk 1 above)
Comment (dsaxton) [reply pandas/tests/groupby/test_function.py:1656]: Which part seems incorrect? The dtype of the index being object is maybe odd, but otherwise seems okay?

Comment (jreback) [reply pandas/tests/groupby/test_function.py:1656]: it seems your comment about is not correct, e.g. we have NA in all cases. is that not what you meant here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_function.py (hunk 1 above)
Comment (dsaxton) [reply pandas/tests/groupby/test_function.py:1656]: I think that comment was from before adding the conversion using maybe_cast_result (so the original test had NaN for Series but NA for DataFrame): https://github.com/pandas-dev/pandas/pull/32914/commits/2a3f8145769d055886a48bc6b9c6dc4e7ed257e3

Comment (jreback) [reply pandas/tests/groupby/test_function.py:1656]: ok this is fine then

**Review by jreback — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
> > **if** this issue **does** affect other EAs, then this PR is a candidate for backport, so would rather keep the changes here minimal.
> 
> It doesn't affect others that I'm aware of (there's a similar but as far as I can tell unrelated problem with boolean data type which I created a separate issue for)

there is a problem also with timedelta, see https://github.com/pandas-dev/pandas/issues/34051#issuecomment-626688541, but this also occurs on 0.25.3

so won't backport for now.
